### ISSUE TITLE
A shortcut for accessing StreamField blocks by name

### ISSUE THREAD

**Issue by gasman** [wagtail/wagtail#7240]
### Is your proposal related to a problem?

It's common for people to want to retrieve a block of a particular type from a StreamField value - for example, if 'hero image' is one of the block types defined in a page's content, a developer may want to pull out the first such hero image for use in listings or as a social share image. ([An example of this in the wild](https://stackoverflow.com/questions/67855617/cannot-get-blog-listing-using-get-context-to-work-properly/67860016#67860016)...)

This can be done in user code (either in Python or Django template code) by looping over the StreamField value until the appropriate block is found - however, I think this is a common enough scenario that Wagtail should provide a standard method / idiom for it.

### Describe the solution you'd like

The `StreamValue` class could implement a `by_type` method that returns a list of blocks matching the given type:

    page.body.by_type('hero_image')  # returns a list of all `hero_image` blocks in the `body` StreamField

Passing a parameter to `by_type` in this way would not be possible within a template, so as an additional feature, `by_type` could return a (lazily-evaluated?) dict consisting of all blocks organised by type when invoked with no parameters. This would make it possible to write something like:

    <h1>{{ page.body.by_type.heading.0 }}</h1>

The `by_type` method will need tests and documentation.

### Describe alternatives you've considered

StreamValue could be updated to behave as something like a dict as well as a list, e.g.`page.body['hero_image']` - however, this feels a bit ambiguous about whether it should return a single item or a list. (Also, in template code where `foo.bar` is equivalent to `foo['bar']`, it means the block names end up in the same namespace as methods on StreamValue, which feels icky.)

Given that people will probably be using this for retrieving a single item most of the time, we could borrow BeautifulSoup's pattern of `page.body.find('hero_image')` (or `page.body.find(type='hero_image')`?) which returns the first item or None, and `page.body.find_all('hero_image')` which returns a list. It's not so obvious how to make this syntax work nicely within templates, though.

### Additional context

Perhaps something for a later phase, but it's worth bearing in mind that for this kind of StreamField access, optimisations like `bulk_to_python` and #7239 are actually counter-productive, since they're pre-fetching all database objects in the stream but we only want one of them. Maybe this calls for us to treat StreamValues more like querysets, where the actual data retrieval happens as late as possible and these kinds of 'filter' operations act as modifiers to that final lookup.

**Comment by Scotchester** [comment]
This seems like a great idea! What do you think about calling it `blocks_by_type`? Just `by_type` feels a little ambiguous to me.

**Comment by Tijani-Dia** [comment]
I tried to address this.

> by_type could return a (lazily-evaluated?) dict
I don't get this point however. Is it related to the fact that a block can exist in `_raw_data` and not in `_bound_blocks`?

**Comment by gasman** [comment]
@Tijani-Dia Yes, it's related to that. If one of the blocks in the stream is really expensive to evaluate (such as an image gallery containing 100 images), then a simple implementation of `blocks_by_type` that loops over all blocks to build a dict will end up doing that evaluation, even if the caller is doing something like `page.body.blocks_by_type()['heading']` which never looks at that block. Instead, `blocks_by_type()` could return a custom object that behaves like a dict, but only starts scanning through the blocks when we know which type we're looking for.

Don't worry about this too much... I suspect it'll add quite a lot of complexity, and it can always be done as a future optimisation.

**Comment by chosak** [comment]
So the StreamField documentation uses both "name" and "type" when referring to the key used to identify blocks.

[From the docs](https://docs.wagtail.io/en/stable/topics/streamfield.html):

> The block types available to authors are defined as a list of `(name, block_type)` tuples: ‘name’ is used to identify the block type within templates, and should follow the standard Python conventions for variable names: lower-case and underscores, no spaces.

What can be confusing is that this name/type is distinct from what developers sometimes think of when they think of "block type" (or at least what I think of), which is _the class name of the subclass of a `StructBlock`_. For example, consider this block that might provide a common way to describe a person:

```py
class PersonBlock(blocks.StructBlock):
    name = blocks.CharBlock()
    email = blocks.EmailBlock()
    photo = ImageChooserBlock(required=False)
```

Colloquially I would bet developers would refer to _this_ as the block's "type". But this block might be used in different ways on different page models, necessitating different "types":

```py
class BlogPage(Page):
    body = StreamField([
        ('author', PersonBlock()),
    ])

class PublicationPage(Page):
    body = StreamField([
        ('primary_author', PersonBlock()),
        ('coauthor', PersonBlock()),
    ])
```

Note that three different "types" have been used for the same block: `author`, `primary_author`, and `coauthor`. Given that the developer for each page model is going to define fields the way that make sense for that page, there isn't necessarily an easy way to ensure that these always match.

(In the latter example above, it also seems like it'd be useful to provide the ability to lookup by `PersonBlock` regardless of the top-level "type".)

When using `StructBlock` subclasses you can also end up with code that contains duplicative information, e.g.

```py
class BlogPage(Page):
    body = StreamField([
        ('author', AuthorBlock()),
        ('content', ContentBlock()),
        ('photo', PhotoBlock()),
    ])
``` 

in which case the value/meaning of the "type" feels a little bit extraneous.

Granted, this only applies to `StructBlock` subclasses, and not when built-in blocks are used at the top level of StreamField. An all this being said, I realize this gripe is pretty intrinsic to the way that StreamField is defined, and none of it takes away from the value proposed here or implemented nicely by @Tijani-Dia in #7266.

It _does_ make me wonder whether it's at least considering the method name `blocks_by_name` instead of `blocks_by_type`, to disambiguate between potential conflicts with how people think about block "types". Thinking about this I'm still pretty sure that `blocks_by_type` makes the most sense but I wanted to just raise the point to see what you thought @gasman.

**Comment by Tijani-Dia** [comment]
> Don't worry about this too much... I suspect it'll add quite a lot of complexity, and it can always be done as a future optimisation.

Alright @gasman, even if I have to admit that I'm worrying about this lol.

@chosak Thanks for your great feedback. Indeed I didn't make much difference between block type and block name. thanks for your enlightenment on this.  I also think that renaming `blocks_by_type` to `blocks_by_name` will be less confusing. 

> I wanted to just raise the point to see what you thought @gasman.

I'm waiting to see to

**Comment by gasman** [comment]
@chosak That sounds fair. The distinction between 'types' and 'names' is a bit blurred, partly due to the detail that when you write something like `blocks.CharBlock()` in a block definition, you're instantiating CharBlock even though you're not notionally creating any objects, just writing a definition. And that's not something we're going to solve here...

I'm happy to go with `blocks_by_name` as the less ambiguous choice. The one snag is that - following the names from @Tijani-Dia's PR - `blocks_of_name` and `first_block_of_name` don't sound right, because blocks don't "belong to" a name. So maybe the answer is to overload the methods, depending on whether you pass a block name or not:

* `blocks_by_name('heading')` - returns a list of all `heading` blocks
* `blocks_by_name()` - returns a dict mapping block names to the list of blocks for that name, so that you can write things like `{% for block in page.body.blocks_by_name.heading %}` in templates
* `first_block_by_name('heading')` - returns the first `heading` block, or None if no blocks of that type exist

and maybe (for symmetry)...

* `first_block_by_name()` - returns a dict mapping block names to the first block for that name (or None if no such blocks exist), so that you can write `{{ page.body.first_block_by_name.heading }}`

**Comment by yash112-lang** [comment]
Hey @zerolab, is this issue is still open, can I work on it.

**Comment by zerolab** [comment]
[yash112-lang](https://github.com/yash112-lang) we do not assign issues to anyone outside of the core team.

Please read https://github.com/wagtail/wagtail/wiki/New-contributor-FAQ first. 

This issue has a potential fix at #7266 which needs (re-)review.

**Comment by lb-** [comment]
Implemented via https://github.com/wagtail/wagtail/pull/8735  - 8cdb3f3a60d55a92cff470841205ce7d55523df9
